set up SIG-etcd (#7372)
kubernetes/community: sig-etcd/charter.md

# SIG etcd Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [sig-governance].

[Kubernetes Charter README]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/README.md
[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md

## Scope

> Owns the etcd project and how it is used by Kubernetes.

Reviewer: note that you gain ownership of the current, very cumbersome and undeterministic process of updating etcd server/client in k8s.

this is currently a best effort from community members and issues and PRs just run stale:
https://github.com/kubernetes/kubernetes/issues/117648
https://github.com/kubernetes/kubernetes/issues?q=is%3Aopen+label%3Aarea%2Fkubeadm+etcd

various product tooling allows

Reviewer 2: Historically it were etcd maintainers who bumped the etcd image in Kubernetes. Reason was that Kubernetes scalability tests are the major signal in etcd qualification, so minor bumps etcd version were done immediately. 

What's more cumbersome is security patching, which is a problem because there are the etcd k8s image is also used by kubeadm. With the SIG in place, I think we can discuss changing/improving the release process.

Reviewer: > Historically it were etcd maintainers who bumped the etcd image in Kubernetes. Reason was that Kubernetes scalability tests are the major signal in etcd qualification, so minor bumps etcd version were done immediately.

i think i vaguelly recall this period.

> What's more cumbersome is security patching, which is a problem because there are the etcd k8s image is also used by kubeadm. With the SIG in place,

the biggest current pain points for etcd update in k8s are:
- undocumented, manual image promotion process.
- no dedicated top level approver to own etcd k/k updates as it touches a lot of things.
- too many steps, dufferent repos, client vs server updates

kubeadm updates are simple - changing a small version map / constant. by updating kubeadm k8s+etcd gains upgrade signal from etcd version at k8s N-1 to etcd version at k8s N. the kube-up upgrade suite is still not working, IIRC.

>  I think we can discuss changing/improving the release process.

happy to participate.

- Maintenance of [etcd image] packaged with Kubernetes

[etcd]: https://github.com/etcd-io/etcd
[etcd-io organization]: https://github.com/etcd-io
[etcd image]: https://github.com/kubernetes/kubernetes/tree/master/cluster/images/etcd

#### Cross-cutting and Externally Facing Processes

- Specifying, testing and improving the implicit Kubernetes-ETCD Contract, which includes storage requirements, write and delete requirements, read requirements and watch requirements.
- Release process of etcd and other binaries belonging to [etcd-io organization]

### Out of scope

- Structure of data stored in etcd by Kubernetes components is owned by SIG API Machinery

## Roles and Organization Management

This SIG follows the Roles and Organization Management outlined in [sig-governance]
and opts-in to updates and modifications to [sig-governance].

### Additional responsibilities of Tech Leads

- Release of etcd and other binaries belonging to [etcd-io organization]

### Deviations from [sig-governance]

- SIG etcd's participation in the Kubernetes release cycle is limited by etcd having a different schedule for its releases.

> - SIG etcd communication utilizes pre-existing forums for communication:

Reviewer 3: I noticed a google meet link instead of Zoom too?

Reviewer 4: Ah, good catch, we are actually using zoom now. Updated the meeting link.

# SIG etcd Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [sig-governance].

[Kubernetes Charter README]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/README.md
[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md

## Scope

Owns the etcd project and how it is used by Kubernetes.

### In scope

#### Code, Binaries and Services

- Development of [etcd] and other repositories under [etcd-io organization]
- Maintenance of [etcd image] packaged with Kubernetes

[etcd]: https://github.com/etcd-io/etcd
[etcd-io organization]: https://github.com/etcd-io
[etcd image]: https://github.com/kubernetes/kubernetes/tree/master/cluster/images/etcd

#### Cross-cutting and Externally Facing Processes

- Specifying, testing and improving [The Implicit Kubernetes-ETCD Contract]
- Release process of etcd and other binaries belonging to [etcd-io organization]

> [The Implicit Kubernetes-ETCD Contract]: https://docs.google.com/document/d/1NUZDiJeiIH5vo_FMaTWf0JtrQKCx0kpEaIIuPoj9P6A/edit?usp=sharing

Reviewer 3: As this document is referenced in the charter, it shouldn't really be something that isn't version controlled (an edit to the google doc, could mean a possible change in interpretation to the charter). Is it possible to either pull out the things that are relevant to the charter with a reference to the original source, or can we pull this document into something in a git repo that is version controlled?

Reviewer 4: Google doc is versioned but only editor could see the versioning history. Git repo is a good idea. @logicalhan @jpbetz @serathius What do you think? How about we convert this doc to a md doc and put in maybe https://github.com/etcd-io/etcd/tree/main/Documentation/contributor-guide?

Reviewer 5: I would suggest just putting a brief description of what that is here for the purposes of stating "this is a process SIG etcd will own", and if we link out at all, to list it under the subprojects instead.

Reviewer 4: Or we could use https://github.com/etcd-io/etcd/issues/15820? This is versioned and easy to track changed, and is in the subproject. wdyt? @cblecker @BenTheElder

Reviewer 4: > I would suggest just putting a brief description of what that is here for the purposes of stating "this is a process SIG etcd will own", and if we link out at all, to list it under the subprojects instead.

@BenTheElder I have updated the charter md to have a statement saying that sig-etcd owns the implicit kubernetes-etcd contract, instead of a linked doc or github issue, ptal. Thanks!

### Out of scope

> - Structure of data stored in etcd is owned by SIG API Machinery

Reviewer 3: ```suggestion
- Structure of data stored in etcd by Kubernetes components is owned by SIG API Machinery
```

## Roles and Organization Management

> This SIG follows adheres to the Roles and Organization Management outlined in [sig-governance]

Reviewer 6: ```suggestion
This SIG follows the Roles and Organization Management outlined in [sig-governance]
```

and opts-in to updates and modifications to [sig-governance].

### Additional responsibilities of Tech Leads

- Release of etcd and other binaries belonging to [etcd-io organization]

### Deviations from [sig-governance]

> - SIG-Etcd's participation in the Kubernetes release cycle is limited by Etcd having a different schedule for its releases.  Particularly, Etcd may or may not participate in the Enhancements process depending on whether an upcoming Kubernetes release targets specific new Etcd features.

Reviewer 3: Is participation in some of our processes (like KEPs) something that sig-etcd may look to do in the future? While it makes sense that etcd would have a different release cycle from the core Kubernetes itself (many subprojects do this too) there might be some benefit to looking at the processes like KEPs for things like PRRs

Reviewer 4: Yes, we do intend to evaluate and adopt other kubernetes processes like PRRs in future. We have a doc to track the things that we will consider along the way: https://tinyurl.com/etcd-sig-checklist.

Reviewer 7: @wenjiaswe Can we also make this info explicit here in the charter?

It's okay for etcd to not follow the Kubernetes release cycle right away. I think we have a few components which don't align to the Kubernetes release cycle. Participation in the KEP process doesn't necessarily imply adhering to the Kubernetes release cycle.

Reviewer 5: Yes, following the release cycle is largely not a requirement for KEPs. Some of the review work like PRR may currently be influenced by the release cycle for practical reasons but etcd does not need to adhere to the kubernetes release lifecycle, KEPs or otherwise.

Most SIGs don't state in their charter which changes will involve KEPs and generally large or controversial changes trigger a request for a KEP by the approvers / OWNERS of projects, at their discretion.

So I don't think that needs determining here.

Reviewer 4: Thank you for your input @BenTheElder @palnabarun . I agree that, even though `Etcd may or may not participate in the Enhancements process` is a valid statement and align with the rest of the SIGs, but `depending on whether an upcoming Kubernetes release targets specific new Etcd features` may cause some confusion, as SIG shouldn't have to adhere to the k8s release cycle to participate in the KEP process.

How about I just remove this statement to align with the charter of other sigs?

- SIG etcd communication utilizes pre-existing forums for communication:
  - Email: [etcd-dev](https://groups.google.com/forum/?hl=en#!forum/etcd-dev).
  - Slack: [#etcd](https://kubernetes.slack.com/messages/C3HD8ARJ5/details/) channel on Kubernetes.
- SIG etcd contributing instructions ([CONTRIBUTING.md]) be defined in etcd project.

[CONTRIBUTING.md]: https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md

### Deviations from [kubernetes-repositories]

- SIG etcd repositories live in github.com/etcd-io
- SIG etcd repositories should (but not must) adopt merge bot, Kubernetes PR commands/bot, OWNERS file.

> - SIG etcd repositories will follow [rules for donated repositories].

Reviewer 8: posting here for convenience for other reviews:

### Rules for new repositories

   * For now all repos will live in `github.com/kubernetes-sigs/\<project-name\>`.
   * Must contain the topic for the sponsoring SIG - e.g.
     `k8s-sig-api-machinery`.  (Added through the *Manage topics* link on the
repo page.)
   * Must adopt the Kubernetes Code of Conduct
   * All code projects use the Apache License version 2.0. Documentation
     repositories must use the Creative Commons License version 4.0.
   * Must adopt the CNCF CLA bot, merge bot and Kubernetes PR commands/bots.
   * All OWNERS of the project must also be active SIG members.
   * Must be approved by the process spelled out in the SIG's charter and a
   publicly linkable written decision should be available for the same.
   * SIG must already have identified all of their existing subprojects and
     code, with valid OWNERS files, in
[`sigs.yaml`](https://github.com/kubernetes/community/blob/master/sigs.yaml)

### Rules for donated repositories

The `kubernetes-sigs` organization is primarily intended to house net-new
projects originally created in that organization. However, projects that a SIG
adopts may also be donated.

In addition to the requirements for new repositories, donated repositories must
demonstrate that:

   * All contributors must have signed the [CNCF Individual
     CLA](https://github.com/cncf/cla/blob/master/individual-cla.pdf) or [CNCF
Corporate CLA](https://github.com/cncf/cla/blob/master/corporate-cla.pdf)
   * If (a) contributor(s) have not signed the CLA and could not be reached, a
     NOTICE file should be added referencing section 7 of the CLA with a list of
the developers who could not be reached
   * Licenses of dependencies are acceptable; project owners can ping
     [caniszczyk](https://github.com/caniszczyk) for review of third party deps
   * Boilerplate text across all files should attribute copyright as follows:
     `"Copyright <Project Authors>"` if no CLA was in place prior to donation
   * Additions of [the standard Kubernetes header](https://git.k8s.io/kubernetes/hack/boilerplate/boilerplate.go.txt)
     to code created by the contributors can occur post-transfer, but should
     ideally occur shortly thereafter.
   * Should contain template files as per the
     [kubernetes-template-project](https://github.com/kubernetes/kubernetes-template-project).

Note that copyright notices should only be modified or removed by the people or
organizations named in the notice. See [the FAQ below](#faq) for more information
regarding copyrights and copyright notices.

### Out of scope

- Structure of data stored in etcd is owned by SIG API Machinery

## Roles and Organization Management

This SIG follows adheres to the Roles and Organization Management outlined in [sig-governance]
and opts-in to updates and modifications to [sig-governance].

### Additional responsibilities of Tech Leads

- Release of etcd and other binaries belonging to [etcd-io organization]

### Deviations from [sig-governance]

- SIG-Etcd's participation in the Kubernetes release cycle is limited by Etcd having a different schedule for its releases.  Particularly, Etcd may or may not participate in the Enhancements process depending on whether an upcoming Kubernetes release targets specific new Etcd features.
- SIG etcd communication utilizes pre-existing forums for communication:
  - Email: [etcd-dev](https://groups.google.com/forum/?hl=en#!forum/etcd-dev).
  - Slack: [#etcd](https://kubernetes.slack.com/messages/C3HD8ARJ5/details/) channel on Kubernetes.
- SIG etcd contributing instructions ([CONTRIBUTING.md]) be defined in etcd project.

> - KEPs created by SIG etcd will not follow Kubernetes release cycle and freezes due.

Reviewer 8: FWIW - on the KEP stuff, we do have a good amount of prior art of KEPs that don't align with the release cycle. If they DO align with it, or are involved with it - thats really what the language in the sig-governance is about (we want the sigs to be responsive)

It might be worth steering to make that clear in the governance docs.

Reviewer 4: Removed.

> Tech Leads must also fulfill all of the responsibilities of the Chair role as outlined in [sig-governance].

Reviewer 8: I don't think this needs to be explicitly called out as a deviation. Quite a few sigs have TLs that are also chairs, we just ask that they be listed under both roles.

Reviewer 4: Removed.

### In scope

#### Code, Binaries and Services

- Development of [etcd] and other repositories under [etcd-io organization]
- Maintenance of [etcd image] packaged with Kubernetes

[etcd]: https://github.com/etcd-io/etcd
[etcd-io organization]: https://github.com/etcd-io
[etcd image]: https://github.com/kubernetes/kubernetes/tree/master/cluster/images/etcd

#### Cross-cutting and Externally Facing Processes

- Specifying, testing and improving [The Implicit Kubernetes-ETCD Contract]
- Release process of etcd and other binaries belonging to [etcd-io organization]

[The Implicit Kubernetes-ETCD Contract]: https://docs.google.com/document/d/1NUZDiJeiIH5vo_FMaTWf0JtrQKCx0kpEaIIuPoj9P6A/edit?usp=sharing

### Out of scope

- Structure of data stored in etcd is owned by SIG API Machinery

## Roles and Organization Management

This SIG follows adheres to the Roles and Organization Management outlined in [sig-governance]
and opts-in to updates and modifications to [sig-governance].

### Additional responsibilities of Tech Leads

- Release of etcd and other binaries belonging to [etcd-io organization]

### Deviations from [sig-governance]

> - SIG etcd does not participate in release planning meetings,
>   retrospectives and burn-down meetings as etcd release process is independent of Kubernetes.

Reviewer 9: ```suggestion
- SIG-Etcd's participation in the Kubernetes release cycle is limited by Etcd having a different schedule for its releases.  Particularly, Etcd may or may not participate in the Enhancements process depending on whether an upcoming Kubernetes release targets specific new Etcd features.
```

Revision based on reality.  We participate in the release process now, given that there's been a lot of coordination around some of the etcd data fixes and Kubernetes updates.  We're not going to reduce participation when becoming a SIG.  This is really not different from subprojects like kubebuilder, Prow, or CRItools.

Reviewer 4: Agree, it makes sense for sig etcd to participate in release planning when a new etcd version will be bumped in the next release.

- SIG etcd communication utilizes pre-existing forums for communication:
  - Email: [etcd-dev](https://groups.google.com/forum/?hl=en#!forum/etcd-dev).
  - Slack: [#etcd](https://kubernetes.slack.com/messages/C3HD8ARJ5/details/) channel on Kubernetes.
- SIG etcd contributing instructions ([CONTRIBUTING.md]) be defined in etcd project.
- KEPs created by SIG etcd will not follow Kubernetes release cycle and freezes due.

Tech Leads must also fulfill all of the responsibilities of the Chair role as outlined in [sig-governance].

[CONTRIBUTING.md]: https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md

### Deviations from [kubernetes-repositories]

- SIG etcd repositories live in github.com/etcd-io

> - SIG etcd repositories should (but not must) adopt merge bot, Kubernetes PR commands/bot, OWNERS file.

Reviewer 9: This will require our contributors to apply for Kubernetes Org membership.  That's probably a good thing overall, but will involve some bootstrapping.

Reviewer 2: That's not true, I expect similar to Kubernetes Org vs Kubernetes-SIGs Org, the membership will be separated from Kubernetes.

Author: It's not, even kubernetes sigs have to apply for membership through org.

Reviewer 8: yup - our orgs all use the same process of applying. Essentially you apply and if approved you can join any of the primary k8s orgs.

I mentioned out of band (https://github.com/kubernetes/community/pull/7372#issuecomment-1633233709) but to directly comment here on some of the hard reqs:

- etcd org is owned and managed by k8s OR repos are transferred to K8s org (NOTE: people can still be granted repo admin rights, its just management of the org)
- membership will be managed by our our automation and applying for org membership will use our process (current etcd org members wouldn't have to apply, we'd directly add)
- EasyCLA will be enabled on the etcd org or any repos transferred to K8s org
- github actions or the CI currently in play can still be used, but we'd still have to run our global stuff (e.g. cncf-cla, lgtm/approve etc.)

Reviewer 4: Thanks for bringing this up, Josh. I have no problem with that.

Reviewer 8: One comment on the owners files, I think those will be a "must" insofar as we can link to the people responsible for the management of the subproject.

Reviewer 4: I am ok with that, do we just need a k8s style OWNERS file in the top level of each subproject under https://github.com/etcd-io?

cc @serathius @ahrtr @spzala @jmhbnz

Reviewer 10: Sounds good to me too!

Reviewer 6: +1 sounds good to me.

Reviewer 4: Removed the "OWNERS file" from deviation, and added an issue in etcd repo to track the effort of adding top level OWNERS file in each subproject repo: https://github.com/etcd-io/etcd/issues/16367
